pyuvdata/uvdata/ms.py

```diff
@@ -55,6 +115,672 @@ class MS(UVData):
+        )
+        int_val = np.finfo(float).max
+
+        if self.multi_phase_center:
+            sou_list = list(self.phase_center_catalog.keys())
+        else:
+            sou_list = [
+                self.object_name,
+            ]
+
+        row_count = 0
+        for sou_name in sou_list:
+            if self.multi_phase_center:
+                sou_ra = self.phase_center_catalog[sou_name]["cat_lon"]
+                sou_dec = self.phase_center_catalog[sou_name]["cat_lat"]
+                pm_ra = self.phase_center_catalog[sou_name].get("cat_pm_ra")
+                pm_dec = self.phase_center_catalog[sou_name].get("cat_pm_dec")
+                if (pm_ra is None) or (pm_dec is None):
+                    pm_ra = 0.0
+                    pm_dec = 0.0
+                sou_id = self.phase_center_catalog[sou_name]["cat_id"]
+            else:
+                sou_ra = self.phase_center_ra
+                sou_dec = self.phase_center_dec
+                pm_ra = 0.0
+                pm_dec = 0.0
+                sou_id = 0
+
+            sou_dir = np.array([sou_ra, sou_dec])
+            pm_dir = np.array([pm_ra, pm_dec])
+            for jdx in range(self.Nspws):
+                source_table.addrows()
+                source_table.putcell("SOURCE_ID", row_count, sou_id)
+                source_table.putcell("TIME", row_count, time_val)
+                source_table.putcell("INTERVAL", row_count, int_val)
+                source_table.putcell("SPECTRAL_WINDOW_ID", row_count, jdx)
+                source_table.putcell("NUM_LINES", row_count, 0)
+                source_table.putcell("NAME", row_count, sou_name)
+                source_table.putcell("DIRECTION", row_count, sou_dir)
+                source_table.putcell("PROPER_MOTION", row_count, pm_dir)
+                row_count += 1
+
+        # We have one final thing we need to do here, which is to link the SOURCE table
+        # to the main table, since it's not included in the list of default subtables.
+        # You are _supposed_ to be able to provide a string, but it looks like that
+        # functionality is actually broken in CASA. Fortunately, supplying the whole
+        # table does seem to work (as least w/ casscore v3.3.1).
+        ms = tables.table(filepath, readonly=False, ack=False)
+        ms.putkeyword("SOURCE", source_table)
+        ms.done()
+
+        source_table.done()
+
+    def _write_ms_pointing(self, filepath):
+        """
+        Write out the pointing information into a CASA table.
+
+        Parameters
+        ----------
+        filepath : str
+            path to MS (without POINTING suffix)
+
+        """
+        if not casa_present:  # pragma: no cover
+            raise ImportError(no_casa_message) from casa_error
+
+        pointing_table = tables.table(
+            filepath + "::POINTING", ack=False, readonly=False
+        )
+        # TODO: we probably have to fix this for steerable dishes
```

Comment:
Is this still a TODO? If so, what line does it refer to?

Reply:
It is, although not quite in the way its written. I think this actually requires some thought on how to implement in UVData -- it basically requires being able to separately handle where the phase center is vs where the primary beam center is. Right now, its set so that the two are the same, so interestingly enough it's fine for steerable dishes, but NOT for something more "PAPER"-like. I'll migrate this down below to where the DIRECTION and TARGET columns get written, since that's the place where things will need to get updated.

Reply:
Ok. I think that currently UVData doesn't have information on where the primary beam center is, just on where the phase center is. You can of course change the phase center of the data set in later processing (although only at the coarseness of the integration). This is an issue in the MWA as well, because it has (electronically) steerable tiles. It would be good to be able to capture that information, we have had some ideas about that in the past (using a different object) that we should talk through.